plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Evaluation/Machine/Cek/Internal.hs

```diff
@@ -199,6 +197,40 @@ newtype ExBudgetMode cost (uni :: GHC.Type -> GHC.Type) fun = ExBudgetMode
     { unExBudgetMode :: forall s. ST s (ExBudgetInfo cost fun s)
     }
 
+{- Note [Cost slippage]
+Tracking the budget usage for every step in the machine adds a lot of overhead. To reduce this,
+we adopt a technique which allows some overshoot of the budget ("slippage"), but only a bounded
+amount.
+
+To do this we:
+- Assume that all machine steps take the same time, so we can track the "pending budget due to
+machine steps" as a simple 'Int'.
+- Actually "spend" the budget when we've done more than some fixed number of steps, or at the end.
+
+This saves a *lot* of time, at the cost of potentially overshooting the budget by slippage*step_cost,
+which is okay so long as we bound the slippage appropriately.
+
+There are two options for how to bound the slippage:
+1. As a fixed number of steps
+2. As a proportion of the overall budget
+
+Option 2 initially seems much better as a bound: if we run N scripts with an overall budget of B, then
+the potential overrun from 1 is N*slippage, whereas the overrun from 2 is B*slippage. That is, 2
+says we always overrun by a fraction of the total amount of time you were expecting, whereas 1 says
+it depends how many scripts you run... so if I send you a lot of small scripts, I could cause a lot
+of overrun.
+
+However, it turns out (empirically) that we can pick a number for 1 that gives us most of the speedup, but such
+that the maximum overrun is negligible (e.g. much smaller than the "startup cost"). So in the end
+we opted for 1, which also happens to be simpler to implement.
+-}
+
+type Slippage = Int
```

Comment:
Can you just directly use `Int#` instead of putting these annoying bangs everywhere?

Reply:
... good point, I probably can!

Reply:
The point where this breaks is `maybeSpendBudget`: the result parameter must be a lifted type, so can't be `Int#`. So then we'd be stuck with lots of `Int`/`Int#` conversions, which seems worse.

Reply:
It could be `withMaybeSpentBudget :: Int# -> (Int# -> CekM s r) -> CekM s r` then, but yeah, probably not worth it and my [other suggestion](https://github.com/input-output-hk/plutus/pull/3210#discussion_r638861605) about how to remove the bangs seems better anyway.

Reply:
I tried to be cleverer and return `(# Int#, CekM s () #)`, but even then you need the bang patterns on the tuple match when you return, annoyingly.